src/lua/trigger.c

```diff
@@ -58,14 +58,18 @@ lbox_trigger_destroy(struct trigger *ptr)
 	if (tarantool_L) {
 		struct lbox_trigger *trigger = (struct lbox_trigger *) ptr;
 		luaL_unref(tarantool_L, LUA_REGISTRYINDEX, trigger->ref);
+		TRASH(trigger);
 	}
+	TRASH(ptr);
 	free(ptr);
 }
 
 static int
 lbox_trigger_run(struct trigger *ptr, void *event)
 {
 	struct lbox_trigger *trigger = (struct lbox_trigger *) ptr;
+	lbox_push_event_f push_event = trigger->push_event;
+	lbox_pop_event_f pop_event = trigger->pop_event;
```

Comment:
Let's make this more clear:

1. Move this right before the point where the trigger might get destroyed, i.e. before luaT_call.
2. Add a comment explaining why you need to do this.
3. Set trigger to NULL afterwards to emphasize that it isn't safe to use it anymore.

```C
int nargs = 0;
if (trigger->push_event != NULL) {
        nargs = trigger->push_event(L, event);
}
/* explain why */
lbox_pop_event_f pop_event = trigger->pop_event;
trigger = NULL;
if (luaT_call(L, nargs, LUA_MULTRET)) {
```